kubric/core.py

```diff
@@ -0,0 +1,212 @@
+      func(owner=self, member=member, frame=frame)
+
+  def __hash__(self):
+    return object.__hash__(self)
+
+  def __eq__(self, other):
+    return id(self) == id(other)
+
+  def __del__(self):
+    for func in self.destruction_callbacks:
+      func(owner=self)
+
+
+# ## ### ####  Materials  #### ### ## #
+
+class Material(Asset):
+  pass
+
+
+class PrincipledBSDFMaterial(Material):
+  color = kt.RGBA(default_value=Color.from_name('white'))
+  metallic = tl.Float(0.)
+  specular = tl.Float(0.5)
+  specular_tint = tl.Float(0.)
+  roughness = tl.Float(0.4)
+  ior = tl.Float(1.45)
+  transmission = tl.Float(0)
+  transmission_roughness = tl.Float(0)
+  emission = kt.RGBA(default_value=Color.from_name('black'))
+
+
+class MeshChromeMaterial(Material):
+  color = kt.RGBA(default_value=Color.from_name('white'))
+  roughness = tl.Float(0.4)
+
+# ## ### ####  3D Objects  #### ### ## #
+
+
+class Object3D(Asset):
+  position = kt.Vector3D(default_value=(0., 0., 0.))
+  quaternion = kt.Quaternion()
+  scale = kt.Vector3D(default_value=(1., 1., 1.))
+
+  up = tl.CaselessStrEnum(['X', 'Y', 'Z', '-X', '-Y', '-Z'], default_value='Y')
+  front = tl.CaselessStrEnum(['X', 'Y', 'Z', '-X', '-Y', '-Z'], default_value='-Z')
+
+  def look_at(self, target):
+    direction = mathutils.Vector(target) - mathutils.Vector(self.position)
+    self.quaternion = direction.to_track_quat(self.front.upper(), self.up.upper())
+
+
+class PhysicalObject(Object3D):
+  velocity = kt.Vector3D(default_value=(0., 0., 0.))
+  angular_velocity = kt.Vector3D(default_value=(0., 0., 0.))
+
+  static = tl.Bool(False)
+  mass = tl.Float(1.0)
+  friction = tl.Float(0.0)
+  restitution = tl.Float(0.5)
+
+  bounds = tl.Tuple(kt.Vector3D(), kt.Vector3D(),
+                    default_value=((0., 0., 0.), (0., 0, 0.)))
+
+  @validate('mass')
+  def _valid_mass(self, proposal):
+    mass = proposal['value']
+    if mass < 0:
+      raise tl.TraitError(f'mass cannot be negative ({mass})')
+    return mass
+
+  @validate('friction')
+  def _valid_friction(self, proposal):
+    friction = proposal['value']
+    if friction < 0:
+      raise tl.TraitError(f'friction cannot be negative ({friction})')
+    return friction
+
+  @validate('friction')
+  def _valid_friction(self, proposal):
+    friction = proposal['value']
+    if friction < 0:
+      raise tl.TraitError(f'friction cannot be negative ({friction})')
+    return friction
+
+  @validate('bounds')
+  def _valid_bounds(self, proposal):
+    lower, upper = proposal['value']
+    for l, u in zip(lower, upper):
+      if l > u:
+        raise tl.TraitError(f'lower bound cannot be larger than upper bound ({lower} !<= {upper})')
+    return lower, upper
+
+
+class FileBasedObject(PhysicalObject):
+  asset_id = tl.Unicode()
+
+  simulation_filename = tl.Unicode()   # TODO: use pathlib.Path instead
+  render_filename = tl.Unicode()       # TODO: use pathlib.Path instead
+
+  material = tl.Instance(Material, allow_none=True, default_value=None)
```

Comment:
What's the default material then? (if you allow none)

Reply:
Depends. I added this option because for katamari we do not want to overwrite the default material imported from file. 
The cleanest solution there would be to construct a `core.Material` when loading them, but that is difficult and might not even be possible in the general case.